doc/src/modules/solvers/solveset.rst

```diff
@@ -1,12 +1,463 @@
+  representing solutions as dictionary we can define a custom add operation
+  where value of `x` variable is added with value of other `x` variable and
+  value of `y` variable with the value of other `y` variable. But what if the user
+  tries to add solution of two equations which have different sets of variables?
+  Say one has `x` and `y` as variables and other has `u` and `v`? We cannot allow this
+  addition because the keys of dictionaries are unordered but this behavior will
+  be inconsistent with the one variable case because it will be absurd to not to
+  allow addition of two numbers, say one obtained by solution of `x = 1` and
+  other by solving `y = 2`.
+
+
+
+Input API of `Solveset`
+-----------------------
+
+``solveset`` has a cleaner input API, unlike ``solve``, It takes a maximum
+of three arguments:
+
+* Equation(s)
+  The equation(s) to solve.
+
+* Variable(s)
+  The variable(s) for which the equation is to be solved.
+
+* domain
+  The domain in which the equation is to be solved.
+
+
+ ``solveset`` removes the ``flags`` argument of ``solve``, which had made
+ the input API messy and output API inconsistent in ``solve``.
+
+
+What is this domain argument about?
+-----------------------------------
+
+ Solveset is designed to be independent of the assumptions on the
+ variable being solved for and instead, uses the ``domain`` argument to
+ decide the solver to dispatch the equation to, namely ``solveset\_real``
+ or ``solveset\_complex``. It's unlike the old ``solve`` which considers the
+ assumption on the variable.
+
+    >>> from sympy.solvers.solveset import solveset
+    >>> from sympy.abc import x
+    >>> from sympy import S
+    >>> solveset(x**2 + 1, x) # domain=S.Complexes is default
+    {-I, I}
+    >>> solveset(x**2 + 1, x, domain=S.Reals)
+    EmptySet()
+
+
+What are the general methods employed by solveset to solve an equation?
+-----------------------------------------------------------------------
+
+ Solveset uses various methods to solve an equation, here is a brief overview
+ of the methodology:
+
+ * The ``domain`` argument is first considered to know the domain in which
+   the user is interested to get the solution.
+
+ * If the given function is a relational (`>=`, `<=`, `>`, `<`), and the
+   domain is real, then `solve\_univariate\_inequality` and solutions are
+   returned. Solving for complex inequalities are not supported yet.
+
+
+ * Based on the ``domain``, the equation is dispatched to one of the two
+   functions ``solveset\_real`` or ``solveset\_complex``, which solves the
+   given equation in complex and real domain respectively.
+
+ * If the given function (equation) is a product of two or more functions,
+   like say `f = g*h`, then the solution to the given equation is the Union
+   of the solution of the equations `g = 0` and `h = 0`, if and only if both
+   `g` and `h` are finite for a finite input. So, the solution is build up
```

Comment:
> What does this finite condition mean?

If `f(x)` and `g(x)` are both finite we can say that the solution of `f(x)*g(x) == 0` is same as `Union(f(x) == 0, g(x) == 0)`, but it is not true in general. `g(x)` can grow to infinitely large for the values where `f(x) == 0`. To be sure that we are not silently allowing any wrong solutions we are using this technique only if both `f` and `g` are finite for a finite input.

The function `_is_finite_with_finite_vars(m)` in `solvers/solveset.py` does this job.